File: src/lucy/transport/schema.py
"""Control-channel wire schema, version 1 (ADR 0011).

Every control message is a flat JSON object = the :class:`Envelope` header
(``v``, ``type``, ``session_id``, ``turn_id``, ``seq``, ``ts_ms``) merged with a
typed payload for its ``type``. No message ever carries audio bytes (ADR 0004).
All models are ``extra="forbid"``: an unknown field is a hard error, so the
schema is the locked contract every transport adapter and the Rust gateway
build against.
"""

from __future__ import annotations

from typing import Annotated, Dict, List, Literal, NamedTuple, Optional, Type, Union

from pydantic import AfterValidator, BaseModel, ConfigDict, Field

from lucy.limits import MAX_CONTROL_DURATION_MS, MAX_CONTROL_TIMESTAMP_MS

ControlTimestamp = Annotated[int, Field(strict=True, ge=0, le=MAX_CONTROL_TIMESTAMP_MS)]
ControlDuration = Annotated[int, Field(strict=True, ge=0, le=MAX_CONTROL_DURATION_MS)]


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Envelope(_Strict):
    v: int = 1
    type: str
    session_id: str
    turn_id: Optional[str] = None
    seq: int
    ts_ms: ControlTimestamp


ENVELOPE_FIELDS = frozenset(Envelope.model_fields)


# -- upstream payloads (gateway -> python) -----------------------------------


class SessionStarted(_Strict):
    transport: str
    caller: str
    codecs: List[str]
    features: List[str] = Field(default_factory=list)


class VadSpeechStart(_Strict):
    at_ms: ControlTimestamp


class VadSpeechEnd(_Strict):
    at_ms: ControlTimestamp
    speech_ms: ControlDuration


class SttPartial(_Strict):
    text: str
    stability: float = Field(ge=0.0, le=1.0)
    provider: str


class SttFinal(_Strict):
    text: str
    provider: str
    stt_ms: ControlDuration


class Dtmf(_Strict):
    digit: str


class AmdResult(_Strict):
    outcome: Literal["human", "machine", "unknown"]
    confidence: float = Field(ge=0.0, le=1.0)


class TtsPlayback(_Strict):
    utterance_id: str
    state: Literal["started", "mark", "finished", "flushed"]
    mark_chars: int = 0


class BargeIn(_Strict):
    at_ms: ControlTimestamp
    during: Literal["speaking", "thinking"]
    utterance_id: Optional[str] = None


class TransportMetrics(_Strict):
    jitter_ms: float
    rtt_ms: float
    packet_loss: float


class SessionEnded(_Strict):
    reason: str


RECORDING_FEATURE = "recording"
RecordingLeg = Literal["caller", "agent", "mixed"]
OPAQUE_RECORDING_REF_PATTERN = r"^[A-Za-z0-9._:-]+$"
OPAQUE_RECORDING_REF_MAX_LENGTH = 128
PHONE_LIKE_RECORDING_REF_MIN_DIGITS = 8


def validate_opaque_recording_ref(value: str) -> str:
    if sum(character.isdigit() for character in value) >= (
        PHONE_LIKE_RECORDING_REF_MIN_DIGITS
    ):
        raise ValueError("recording references cannot contain phone-like values")
    return value


OpaqueRecordingRef = Annotated[
    str,
    Field(
        min_length=1,
        max_length=OPAQUE_RECORDING_REF_MAX_LENGTH,
        pattern=OPAQUE_RECORDING_REF_PATTERN,
    ),
    AfterValidator(validate_opaque_recording_ref),
]
RecordingContainer = Annotated[
    str, Field(min_length=1, max_length=32, pattern=r"^[a-z0-9][a-z0-9._-]*$")
]


class RecordingStarted(_Strict):
    recording_id: OpaqueRecordingRef
    leg: RecordingLeg
    blob_id: OpaqueRecordingRef
    consent_ref: OpaqueRecordingRef


class RecordingUploaded(_Strict):
    recording_id: OpaqueRecordingRef
    leg: RecordingLeg
    blob_id: OpaqueRecordingRef
    upload_url_ref: OpaqueRecordingRef
    duration_ms: int = Field(ge=0)
    byte_count: int = Field(ge=0)
    sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    container: RecordingContainer
    consent_ref: OpaqueRecordingRef


class RecordingFailed(_Strict):
    recording_id: OpaqueRecordingRef
    error_code: str
    retryable: bool = False


# -- downstream payloads (python -> gateway) ---------------------------------


class SessionConfigure(_Strict):
    stt: str
    tts: str
    vad: str


class TtsSpeak(_Strict):
    utterance_id: str
    text: str
    flush: bool = False


class TtsCancel(_Strict):
    utterance_id: str  # a specific utterance, or the literal "all"


class TtsStreamEnd(_Strict):
    """No more ``tts.speak`` directives follow for the current turn - the
    agent's speech stream is complete or was cancelled. Sent exactly once per
    turn; the dev gateway uses it as the deterministic turn barrier when
    echoing playback (card 64). A real media plane derives the same boundary
    from its own playout queue."""


class RealtimeConnect(_Strict):
    provider: str
    model: str


class RealtimeToolResult(_Strict):
    call_id: str
    output_json: str


class DtmfSend(_Strict):
    digits: str


class Transfer(_Strict):
    target: str


class Dial(_Strict):
    target: str
    caller_id: str
    timeout_ms: int


class Hold(_Strict):
    state: Literal["hold", "resume"]
    music: bool = False


class SessionEnd(_Strict):
    reason: str


class RecordingStart(_Strict):
    recording_id: OpaqueRecordingRef
    leg: RecordingLeg
    blob_id: OpaqueRecordingRef
    upload_url_ref: OpaqueRecordingRef
    container: RecordingContainer
    consent_ref: OpaqueRecordingRef


class RecordingStop(_Strict):
    recording_id: OpaqueRecordingRef


DownstreamDirective = Union[
    SessionConfigure,
    TtsSpeak,
    TtsCancel,
    TtsStreamEnd,
    RealtimeConnect,
    RealtimeToolResult,
    DtmfSend,
    Transfer,
    Dial,
    Hold,
    SessionEnd,
    RecordingStart,
    RecordingStop,
]


UPSTREAM_TYPES: Dict[str, Type[BaseModel]] = {
    "session.started": SessionStarted,
    "vad.speech_start": VadSpeechStart,
    "vad.speech_end": VadSpeechEnd,
    "stt.partial": SttPartial,
    "stt.final": SttFinal,
    "dtmf": Dtmf,
    "amd.result": AmdResult,
    "tts.playback": TtsPlayback,
    "barge_in": BargeIn,
    "transport.metrics": TransportMetrics,
    "session.ended": SessionEnded,
    "recording.started": RecordingStarted,
    "recording.uploaded": RecordingUploaded,
    "recording.failed": RecordingFailed,
}

DOWNSTREAM_TYPES: Dict[str, Type[BaseModel]] = {
    "session.configure": SessionConfigure,
    "tts.speak": TtsSpeak,
    "tts.cancel": TtsCancel,
    "tts.stream_end": TtsStreamEnd,
    "realtime.connect": RealtimeConnect,
    "realtime.tool_result": RealtimeToolResult,
    "dtmf.send": DtmfSend,
    "transfer": Transfer,
    "dial": Dial,
    "hold": Hold,
    "session.end": SessionEnd,
    "recording.start": RecordingStart,
    "recording.stop": RecordingStop,
}
DOWNSTREAM_TYPE_BY_MODEL = {model: name for name, model in DOWNSTREAM_TYPES.items()}

MESSAGE_TYPES: Dict[str, Type[BaseModel]] = {**UPSTREAM_TYPES, **DOWNSTREAM_TYPES}


class UnknownControlMessage(ValueError):
    """Raised when a control message names a ``type`` not in the schema."""


class ControlEvent(NamedTuple):
    envelope: Envelope
    payload: BaseModel


def downstream_type(payload: DownstreamDirective) -> str:
    """Return the registered v1 wire name for a downstream directive."""
    type_name = DOWNSTREAM_TYPE_BY_MODEL.get(type(payload))
    if type_name is None:
        raise TypeError("payload is not a registered downstream directive")
    return type_name


def is_downstream_directive(payload: object) -> bool:
    """Return whether ``payload`` is registered on the v1 downstream wire."""
    return type(payload) in DOWNSTREAM_TYPE_BY_MODEL


def parse_event(raw: Dict[str, object]) -> ControlEvent:
    """Validate a flat control-message dict into its envelope + typed payload.

    Unknown ``type`` raises :class:`UnknownControlMessage`; any field that is
    neither an envelope field nor a payload field of that type is rejected by
    the payload model (``extra="forbid"``).
    """
    envelope = Envelope(**{k: v for k, v in raw.items() if k in ENVELOPE_FIELDS})
    model = MESSAGE_TYPES.get(envelope.type)
    if model is None:
        raise UnknownControlMessage("unknown control message type: %r" % envelope.type)
    payload = model(**{k: v for k, v in raw.items() if k not in ENVELOPE_FIELDS})
    return ControlEvent(envelope, payload)


def to_wire(envelope: Envelope, payload: BaseModel) -> Dict[str, object]:
    """Flatten an envelope + payload back into a single wire dict."""
    return {**envelope.model_dump(), **payload.model_dump()}

File: media-gateway-rust/src/control/schema.rs
use serde::{de, Deserialize, Deserializer, Serialize};
use serde_json::Value;
use std::collections::BTreeSet;

const ENVELOPE_FIELDS: &[&str] = &["v", "type", "session_id", "turn_id", "seq", "ts_ms"];

#[derive(Debug, Clone, Serialize)]
#[serde(transparent)]
pub struct ControlMessage(Value);

impl ControlMessage {
    pub fn message_type(&self) -> &str {
        self.0["type"].as_str().expect("validated control type")
    }

    pub fn as_value(&self) -> &Value {
        &self.0
    }
}

impl<'de> Deserialize<'de> for ControlMessage {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let value = Value::deserialize(deserializer)?;
        validate(&value).map_err(de::Error::custom)?;
        Ok(Self(value))
    }
}

pub fn canonical_json(message: &ControlMessage) -> Result<String, serde_json::Error> {
    serde_json::to_string(&serde_json::to_value(message)?)
}

fn validate(value: &Value) -> Result<(), String> {
    let object = value
        .as_object()
        .ok_or_else(|| "control message must be an object".to_string())?;
    let kind = object
        .get("type")
        .and_then(Value::as_str)
        .ok_or_else(|| "control message type is required".to_string())?;
    let payload = payload_fields(kind).ok_or_else(|| format!("unknown type: {kind}"))?;
    let allowed: BTreeSet<&str> = ENVELOPE_FIELDS
        .iter()
        .copied()
        .chain(payload.iter().copied())
        .collect();
    for field in object.keys() {
        if !allowed.contains(field.as_str()) {
            return Err(format!("unknown field {field} for {kind}"));
        }
    }
    for field in ["type", "session_id", "seq", "ts_ms"] {
        if !object.contains_key(field) {
            return Err(format!("missing required envelope field {field}"));
        }
    }
    for field in required_payload_fields(kind) {
        if !object.contains_key(*field) {
            return Err(format!("missing required field {field} for {kind}"));
        }
    }
    for (field, field_value) in object {
        validate_field_type(kind, field, field_value)?;
    }
    Ok(())
}

fn required_payload_fields(kind: &str) -> &'static [&'static str] {
    match kind {
        "barge_in" => &["at_ms", "during"],
        "hold" => &["state"],
        "recording.failed" => &["recording_id", "error_code"],
        "session.started" => &["transport", "caller", "codecs"],
        "tts.playback" => &["utterance_id", "state"],
        "tts.speak" => &["utterance_id", "text"],
        _ => payload_fields(kind).unwrap_or(&[]),
    }
}

fn validate_field_type(kind: &str, field: &str, value: &Value) -> Result<(), String> {
    let valid = match field {
        "v" | "seq" | "ts_ms" | "at_ms" | "speech_ms" | "stt_ms" | "mark_chars" | "timeout_ms"
        | "duration_ms" | "byte_count" => value.as_u64().is_some(),
        "stability" | "confidence" | "jitter_ms" | "rtt_ms" | "packet_loss" => {
            value.as_f64().is_some()
        }
        "retryable" | "flush" | "music" => value.is_boolean(),
        "codecs" | "features" => value
            .as_array()
            .is_some_and(|items| items.iter().all(Value::is_string)),
        "turn_id" => value.is_null() || value.is_string(),
        "utterance_id" if kind == "barge_in" => value.is_null() || value.is_string(),
        _ => value.is_string(),
    };
    if !valid {
        return Err(format!("invalid field type for {field} in {kind}"));
    }
    match (field, value.as_str()) {
        ("outcome", Some(value)) if !["human", "machine", "unknown"].contains(&value) => {
            Err(format!("invalid outcome: {value}"))
        }
        ("during", Some(value)) if !["speaking", "thinking"].contains(&value) => {
            Err(format!("invalid barge-in phase: {value}"))
        }
        ("leg", Some(value)) if !["caller", "agent", "mixed"].contains(&value) => {
            Err(format!("invalid recording leg: {value}"))
        }
        ("state", Some(value))
            if kind == "tts.playback"
                && !["started", "mark", "finished", "flushed"].contains(&value) =>
        {
            Err(format!("invalid playback state: {value}"))
        }
        ("state", Some(value)) if kind == "hold" && !["hold", "resume"].contains(&value) => {
            Err(format!("invalid hold state: {value}"))
        }
        _ => Ok(()),
    }
}

fn payload_fields(kind: &str) -> Option<&'static [&'static str]> {
    Some(match kind {
        "amd.result" => &["outcome", "confidence"],
        "barge_in" => &["at_ms", "during", "utterance_id"],
        "dial" => &["target", "caller_id", "timeout_ms"],
        "dtmf" => &["digit"],
        "dtmf.send" => &["digits"],
        "hold" => &["state", "music"],
        "realtime.connect" => &["provider", "model"],
        "realtime.tool_result" => &["call_id", "output_json"],
        "recording.failed" => &["recording_id", "error_code", "retryable"],
        "recording.start" => &[
            "recording_id",
            "leg",
            "blob_id",
            "upload_url_ref",
            "container",
            "consent_ref",
        ],
        "recording.started" => &["recording_id", "leg", "blob_id", "consent_ref"],
        "recording.stop" => &["recording_id"],
        "recording.uploaded" => &[
            "recording_id",
            "leg",
            "blob_id",
            "upload_url_ref",
            "duration_ms",
            "byte_count",
            "sha256",
            "container",
            "consent_ref",
        ],
        "session.configure" => &["stt", "tts", "vad"],
        "session.end" | "session.ended" => &["reason"],
        "session.started" => &["transport", "caller", "codecs", "features"],
        "stt.final" => &["text", "provider", "stt_ms"],
        "stt.partial" => &["text", "stability", "provider"],
        "transfer" => &["target"],
        "transport.metrics" => &["jitter_ms", "rtt_ms", "packet_loss"],
        "tts.cancel" => &["utterance_id"],
        "tts.playback" => &["utterance_id", "state", "mark_chars"],
        "tts.speak" => &["utterance_id", "text", "flush"],
        "tts.stream_end" => &[],
        "vad.speech_end" => &["at_ms", "speech_ms"],
        "vad.speech_start" => &["at_ms"],
        _ => return None,
    })
}
